### ISSUE TITLE
Use of Multiple Cores

### ISSUE THREAD

**Issue by RohithCIS** [inkstitch/inkstitch#836]
The Rendering, PDF exports and everything else uses only 1 out of 8 cores of my CPU. It makes it too slow. Can you modify it to use multiple cores or have some steps that I can do to change them on mine?

**Comment by lexelby** [comment]
Definitely something we want to look into.  It's a little tricky because rendering each successive object needs to know where stitching ended on the previous object.  For example, if you have two fill stitch objects, Ink/Stitch figures out where the last stitch in the first one falls and then picks a spot as close as possible to start on the next object.

There's still room for parallelism, but right now we're still in the "get it to work at all" stage.  Optimization is down the road. :)  Of course, if someone wanted to work on it and submit a PR, that would be awesome!

**Comment by lexelby** [comment]
Another thing I want to try to do is have it cache the stitched version of each object so that it doesn't have to rerender if the object doesn't change.

**Comment by RohithCIS** [comment]
> Another thing I want to try to do is have it cache the stitched version of each object so that it doesn't have to rerender if the object doesn't change.

This would be a really nice feature. Let me see if I can contribute also. I will try to work on it in the weekends.

**Comment by kaalleen** [comment]
This is very exciting. Thanks for having a look into it!!

Be aware that we are in a process with a lot of code changes involved. Almost every file will be touched when we will switch from python 2 to python 3, an updated inkex version and also switch the gui from wxpython to electron. Maybe it's worth waiting a little - or just keep it in mind while working on it ;)

**Comment by ecc521** [comment]
@RohithCIS: You probably know this already, however shrinking the SVG images tends to massively speed up the embroidery. I tend to run my simulations in the 1-2 inch range (~2 seconds), then lock scale, and bring it up to full size for the final version, and find waiting ~12 seconds once (just a random 5x6 inch stitch pattern on my machine) an awful lot less annoying than "Error during autofill" (which seems heavily related to extremely small fill, often junk fill generated by something somewhere). Your situation may well be quite different though. How slow is it going for you?

On a related note, is there a good reason why the popup window can't just be spun off from inkscape (for the functions that don't modify inkscape)? Not locking up inkscape would probably be a much easier way to achieve similar-feeling performance gains.

**Comment by RohithCIS** [comment]
@ecc521 I didn't think of that. That's a nice trick. Thank You.

> On a related note, is there a good reason why the popup window can't just be spun off from inkscape (for the functions that don't modify inkscape)? Not locking up inkscape would probably be a much easier way to achieve similar-feeling performance gains.

Yes!! That is how it needs to be. I think the electron implementation will solve this to an extent.

**Comment by lexelby** [comment]
Hmm, hadn't thought of that.  I guess that would only work for Simulate and Print?

It may seem a little weird if you run Simulate and then change the object, and the Simulate window doesn't reflect the changes?

**Comment by ecc521** [comment]
Personally, I wouldn't find it weird, given how long Simulate takes to process - anything that takes more than a few seconds probably won't dynamically update.

That said, a simple message "If you make changes in Inkscape, you will need to run Simulator again to view those changes" would probably be fine.

**Comment by ValiHar** [comment]
Hi @lexelby, has there been any progress on this? The program is great but it is veryyy slow and my hardware should be sufficient. Maybe I can support you in some way? What is the current status of improving performance and stability? :)

**Comment by lexelby** [comment]
The biggest thing we've done recently in the realm of performance is the addition of stitch plan caching.  The most recent version of Ink/Stitch will save the stitch plan for each object and reuse it if the object or its params have not changed.  Note that some objects make use of the position of the last stitch in the object stitched immediately before that object, and so changes to objects early in the stitching order can cause a lot of objects' stitch plans to be recalculated.

But in general, yes, help would be greatly appreciated!  If you can see optimizations or if you'd like to try to tackle multi-core stitch plan calculation, by all means go right ahead.  I'm interested to see if we can speed things up.

In general, our approach has been to make _reasonable_ attempts at optimization while designing new features, but also to prioritize code readability and understandability, sometimes at the expense of speed.  We're still very much in the R&D stage in this project, trying to build enough features to make Ink/Stitch a compelling digitizing platform.  Improving speed hasn't been a huge priority yet.

I"m wondering what kind of performance you're seeing though?  It's normal for some objects, especially complex fills, to take a couple seconds to render.  I don't expect it to take minutes though.  I know in the past we have had elusive font-related problems on Linux that I haven't been able to reproduce or get to the bottom of.  Is this by any chance what you're encountering? https://github.com/inkstitch/inkstitch/issues/575

**Comment by lexelby** [comment]
I took a look at the feasibility of multi-core stitch plan calculation.  There are two hurdles to overcome in order to make this possible, one relatively easy and one that will make our lives quite difficult.

The first is that some objects depend on the position of the previous stitch.  Fill stitch objects set their starting point to a point as close as possible to the last stitch from the previous object, in order to limit the jump stitch length.  They do this unless the user specifies the start position.  This means that we wouldn't be able to parallelize stitching of two objects if the second one is a fill without a start position set.  Not a big deal though -- we'd just need to split the stitch plan into parallelizable groups of objects and farm those out to separate cores.

The bigger problem is how exactly to split work across cores.  Python, of course, has the Global Interpreter Lock, and this is cpu-intensive work, so simple threading won't help.  The alternative is multiprocessing, and that's where we run smack into a roadblock.  How do we transfer a unit of work to one of the child processes?

We'd need to send an EmbroideryElement instance, which contains a reference to an XML node.  However, we'll need access to the full SVG DOM, because we'll need to be able to find commands that apply to this element, plus CSS, plus transforms applied to ancestors of this node.

Worse yet, it seems that lxml doesn't support python's pickling (or rather, unpickled XML node objects throw errors).  This means we'd need to serialize the entire SVG and pass the whole thing to all child processes.  Each process would need to _separately_ re-parse that SVG with lxml (actually `inkex`).  That's gonna take a fair bit of work and will eat into the benefit we get from using multiple cores.

On Linux and MacOS, we can tell multiprocessing to `fork()` so that child processes get a copy of the full memory state of the script.  Windows doesn't have `fork()`, though, so we can't depend on that behavior.  In fact, Windows is a pain for another reason: multiprocessing has to re-execute the whole program, and that gets really tricky with PyInstaller.  It's gonna be a pain to get this all working properly.

Bottom line: multi-core would be tricky and may not actually improve our speed much after all the overhead is taken into account.

**Comment by lexelby** [comment]
Okay, after some testing, it seems as if lxml is _shockingly_ fast, so perhaps there's something to this -- but the multiprocessing juggling is definitely gonna be a pain.

**Comment by lexelby** [comment]
One reason this is going to be really tricky: it takes 1-2 seconds to start a worker process (at least on my laptop).  I suspect this is because of a) the relatively large amount of code in Ink/Stitch and b) our extensive use of decorators, which means a fair bit of code runs at startup.

Do we always start a worker pool?  What if the main process could get everything done before the worker processes could even start up?  That means we take a completely unnecessary 1-2 second penalty.

I guess the best we can do is to start the pool and feed it a no-op async task, then start processing the stitch plan in the main process.  If the async workers return, then we can feed the rest of the tasks to the pool.  Things are starting to get pretty complex.
